### PR TITLE
Database Validation and Cleanup

### PR DESCRIPTION
**Pull request by Ebag333**
This addresses three issues:

1. Character skills exist for a character that doesn't.  See #917
2. Damage pattern on a fit is set to a damage pattern that doesn't exist.  See #777
3. Character on a fit is set to a character that doesn't exist. (Can't find which issue mentioned this.)

@blitzmann I know that you wanted to avoid running database validation and cleanup each time, but I don't think it's avoidable.  We are running into scenarios where people can't even launch Pyfa, so they can't get to a point when they _could_ click a button to fix it.

And yes, I know that the first two issues won't cause that scenario.  But since a lot of users won't ever take the time to post here or on the forums about it (just rage quit and go to another tool), I think that this is the best approach.

Plus, this takes basically nil time.  For a clean DB:
```
2016-12-26 21:39:18,069 root                     DEBUG    Starting database validation.
2016-12-26 21:39:18,072 eos.db.saveddata.databaseRepair DEBUG    Start databsae validation and cleanup.
2016-12-26 21:39:18,075 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for character skills.
2016-12-26 21:39:18,082 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for orphaned characters attached to fits.
2016-12-26 21:39:18,089 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for orphaned damage patterns attached to fits.
2016-12-26 21:39:18,098 root                     DEBUG    Completed database validation.
```

Takes all of 26 milliseconds and likely most of that is from the actual logging.

Ran against my personal DB (where I figured out the uniform damage profile issue), only took literally a handful of milliseconds more to clean up hundreds of records.

Now, this doesn't _fix_ the root issue.  But since we don't know what causes it exactly, and we have people who are currently broken and non-functional, we might as well clean it up to get them back on their feet.

Broke this out into it's own class so we can easily extend it with more checks.

### CODE REVIEW

**Comment by blitzmann** [comment]
Doesn't seem too bad. Eventually I would only want this to run if we ever run into a sql exception, which would require messing with custom exception handlers. Buy for now this should do.

Also, your record counting... there should be a much easier way. Len(results) should work?

**Comment by Ebag333** [comment]
>Also, your record counting... there should be a much easier way. Len(results) should work?

You'd think, wouldn't you?

`{TypeError}object of type 'ResultProxy' has no len()`

Searching for it seems that there's no real good solutions for it.  Some folks write custom functions to do it, but it's more complicated that need be IMHO.

So the various `.count`, `len`, etc don't work for this.  Could convert the object type to a dict, but that'd basically end up being just as complex (really more so).

It's not pretty, but it takes a fraction of a millisecond to do (even if there's hundreds of records), and most times there won't be any records at all.

**Comment by Ebag333** [comment]
Also, FYI, this should be ready to merge.

**Comment by blitzmann** [comment]
Ah, yeah you'd have to do a `.fetchall()` to get the actual results - that should get you either a `len()` or a `rowcount` property maybe.

Or you can do a `SELECT COUNT(*)` into the result and use that to drive whether or not to run the other query.

Or just remove the counting completely - a simple DELETE statement that covers the condition should suffice.

But none of this is breaking and can remain as is - like I said, eventually I want to change how this is triggered, so if anything it can be changed then.

**Comment by Ebag333** [comment]
>Ah, yeah you'd have to do a .fetchall() to get the actual results - that should get you either a len() or a rowcount property maybe.

I can fiddle with that, but my understanding is that the `returnrows` and `rowcount` properties are never populated for a select, so I'm doubtful that'd work.

I haven't tried `.fetchall()` so I can see if that makes a difference.

>Or you can do a SELECT COUNT(*) into the result and use that to drive whether or not to run the other query.

You still end up with a result you have to loop through or otherwise extract the first record, so it's not really much more efficient.  Would be a single loop instead of possible a few hundred, but lines of code wise it's not any different.  It'd basically end up being what I did for getting the ID for Uniform damage profile and the All V character.

>Or just remove the counting completely - a simple DELETE statement that covers the condition should suffice.

This makes me uncomfortable.  I don't like blindly deleting things, it's a bad practice with SQL data and I want it well documented if we *DO* delete anything.

Speaking of which, I'm assuming that logging errors will always output to the log, correct?

>But none of this is breaking and can remain as is - like I said, eventually I want to change how this is triggered, so if anything it can be changed then.

Yeah.  I'm just getting tired of all the new issues being opened with the same root causes.  I just want to get something in place to get this cleaned up so we don't spend our time troubleshooting the same thing over and over again, and we don't have users rage quitting without even telling us (which I'm sure happens).

I think that we should get this implemented and extend it as we figure out other serious Pyfa breaking scenarios.  What little time this costs us starting up can be saved elsewhere (like fixing the triple loop looking for if you have a fit for each ship in each class).

**Comment by blitzmann** [comment]
>  I'm assuming that logging errors will always output to the log, correct?

You assume too much.

Logging in pyfa is hit and miss. I made an attempt once to actually get it to work with errors, but doesn't seem to work all the time (like, 90% of the time it doesn't). This is actually an area that I want to revisit soon and harden up - logging all errors to a file, or if run with `-d` redirect all output to a file.

I first want to get #247 looked at once more and tested on the different platforms - as long as it notifies the user of the exception, logging it to disk is less important.

**Comment by Ebag333** [comment]
>You assume too much.

Let me rephrase.

Info and error is _supposed_ to log to the file.  Debug _should_ only log if `-d` is there.

I know that logging isn't perfect.  @a-tal had a fit when he saw we used the built in logging at all, because it's slow and heavy.  And from debug stepping through it, I can't disagree.

Not sure what we would use to replace it, but even if `error` doesn't properly output now I wanted to get it in place so it's there to properly output once we fix logging and error reporting.

**Comment by blitzmann** [comment]
@Ebag333 made a couple modifications to the PR; see latest commit. I didn't like the idea of SELECT * as that pulls back all data from the table, which could potentially be a performance issue. I find this cleaner. If you don't see anything wrong with these changes, I'll go ahead and merge. :)

**Comment by Ebag333** [comment]
Ran against an empty DB:

```
Traceback (most recent call last):
  File "C:\Program Files (x86)\JetBrains\PyCharm Community Edition 2016.3\helpers\pydev\pydevd.py", line 1596, in <module>
    globals = debugger.run(setup['file'], None, None, is_module)
  File "C:\Program Files (x86)\JetBrains\PyCharm Community Edition 2016.3\helpers\pydev\pydevd.py", line 974, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "./pyfa.py", line 125, in <module>
    import service.prefetch
  File ".\service\prefetch.py", line 68, in <module>
    database_cleanup_instance.OrphanedFitDamagePatterns(eos.db.saveddata_engine)
  File ".\eos\db\saveddata\databaseRepair.py", line 59, in OrphanedFitDamagePatterns
    if results.fetchone()['num'] > 0:
TypeError: 'NoneType' object has no attribute '__getitem__'
```

I'm getting `{TypeError}'NoneType' object has no attribute '__getitem__'` regardless if there's valid data, bad data, or no data at all.

I do see that results is returning a valid count, but I'm not sure that the method you're using to try and extract it works.

I submitted a commit with an alternate way of getting it. it's basically the same method used to grab the IDs (though we don't check if more than 1 record was returned, since we'll only ever get one).

**Comment by blitzmann** [comment]
What do you mean an empty DB? I run this against a brand new saveddata.db and it runs fine. Even if I delete the characters and damage patterns that are created...

After looking at the SQLAlchemy docs, I think there's a better way of doing this anyway - made a commit. Please try to break it again (and explain in more detail how to broke the other one please, because I can't figure it out lol)

**Comment by Ebag333** [comment]
>After looking at the SQLAlchemy docs, I think there's a better way of doing this anyway - made a commit. Please try to break it again (and explain in more detail how to broke the other one please, because I can't figure it out lol)

Uhh, I deleted the DB, ran Pyfa, it broke. vOv

Like I said initially, I tried a BUNCH of ways to get the data out of the results because I wanted the row count. The ONLY way I could get it to work was by looping through the results. Googling and searching through Sqlalchemy docs seemed to confirm that. 

if you found a better way I'm all for that. You must have a slight difference in your environment than I do, maybe a different version of Sqlalchemy or something. But since we have users that will have various versions, the low tech solution (even if it's slightly fugly) is probably the best way to go if it works reliably. 

Anyway, I'll try the latest commit. I'm curious to see how you solved this, because I can think of a good number of scenarios where I would like to query the DB directly, and don't want to have to loop through results just to get a simple count.

**Comment by Ebag333** [comment]
>Uhh, I deleted the DB, ran Pyfa, it broke. vOv

On mobile so can't edit (grrr github) but I also tried it on a DB with data. It just doesn't work in my environment at all.

**Comment by Ebag333** [comment]
>if row and row['num'] > 0:

Random Python thought, but instead of 
`if row and row['num'] > 0:`
Shouldn't we be able to do
`if row:`
since in Python false = None = 0?

Hmm that doesn't necessarily test existence though. I've had problems with that before, but that's not an issue here since we'll always get a number back.

**Comment by Ebag333** [comment]
Okay, so did a few more things.

I added a check and fix for if there's null values (we were only checking missing, which doesn't actually catch null) for both damage pattern and character IDs on fits.

I broke out the query logic (where we're actually looking for results) into it's own function in `queries.py`.  Makes sense since we're already doing it twice, and will likely do it many more times down the road.

The function also converts the resultsproxy into a standard dict because eff that shit.  So now the object that gets returned we can do a simple `__len__()` and hey, there's the number of results we have, how easy is that?  (Technically it's a named dict for each row and the container object is a simple list, seemed the easiest and cleanest way to handle it.)

This function isn't designed to handle everything for you and be very standardized like the other queries (like say `getCrestCharacter`).  This expects you to know what you're doing and be ready to handle the return yourself (so know what the actual table column names are, or specifically alias them in the query).  But it should be good for one off scenarios, like that stupid takes-5-to-60-seconds-to-load loop within a loop within a loop where we're just getting a list of all the ship groups that have a fit in them so we can filter it down.


I did tighten up the `if` statements for the counts as mentioned above, but didn't mess with the queries.  There's no real reason to use the `executeDatabaseQuery` function for that, it'd be heavier than needed, and we couldn't use `__len__()` anyway unless we go back to querying all rows.  Better to keep it at just 2 lines of code.

Anyway, I think we have a winner here.  If you're okay with what I did breaking it out, can ya merge it in?  I want to use that new query function so I can shoot that long load issue in the head. :D
